Adding Apple frameworks > flutter plugins doc

leighajarett opened the pull request:
_Description of what this PR is changing or adding, and why:_ This PR creates a new document in the iOS platform integration section which maps commonly used Apple frameworks to Flutter plugins / dart pacakges. We often see questions like "Can I use XXXKit in a Flutter app?" -- this document seeks to help iOS developers quickly get started with Flutter and answer those types of questions.

_Issues fixed by this PR (if any):_ Fixes https://github.com/flutter/website/issues/7572

## Presubmit checklist
- [x] This PR doesn’t contain automatically generated corrections (Grammarly or similar).
- [x] This PR follows the [Google Developer Documentation Style Guidelines](https://developers.google.com/style) — for example, it doesn’t use _i.e._ or _e.g._, and it avoids _I_ and _we_ (first person).
- [x] This PR uses [semantic line breaks](https://github.com/dart-lang/site-shared/blob/main/doc/writing-for-dart-and-flutter-websites.md#semantic-line-breaks) of 80 characters or fewer.

leighajarett commented:
@jmagman and @hellohuanlin in case you would like to take a look

hellohuanlin commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+framework for pushing user-facing notifications.
+
+In Flutter, you can use the [`firebase_messaging`][] plugin, 
+which uses Firebase Cloud Messaging. 
```
Is there a plugin that supports APNS?

leighajarett replied:
It seems like there is but it's from an unverified publisher so not sure if it's worth mentioning: https://pub.dev/packages/flutter_apns

I haven't personally used FCM before but it seems like you may be able to use APNs through their integration: https://firebase.flutter.dev/docs/messaging/apple-integration/

hellohuanlin replied:
ah gotcha. maybe worth mentioning, since APNS may be what devs are looking for, if they have all their severs setup already.

leighajarett replied:
Good point thanks!

hellohuanlin commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+like the accelerometer or the gyroscope.
+
+In Flutter, to read sensor data, 
+you can use the [`sensors_plus`][] plugin.
```
is this 1P or 3P?

leighajarett replied:
3P -- to me the purpose of this document is not to highlight 1P plugins, it's to highlight plugins that allow you to accomplish the same tasks that you would with Apple's frameworks

hellohuanlin commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+Apple Maps within your app. 
+
+In Flutter, you can embed Google Maps using the 
+[`google_maps_flutter`][] plugin. 
```
iirc there's also a plugin based on mapkit. Should we also mention it?

hellohuanlin replied:
yeah,  probably better not to include it given its current state.

hellohuanlin commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+
+### Persisting data
+In SwiftUI, to persist small pieces of data in your application, 
+you would likely use the `@AppStorage` property wrapper.
```
maybe also mention NSUserDefault

leighajarett replied:
Added!

hellohuanlin commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+to access health and fitness data.
+
+In Flutter, you can use the [`health`][] plugin 
+to access `HealthKit` data.
```
this seems 3P plugin too. should we be careful about endorsing a 3P plugin in our official doc?

sfshaza2 replied:
Yes, correct, we need to be careful.

leighajarett replied:
There are numerous 3P plugins in this doc. Is there a way we can mention them without "endorsing"? Maybe a sentence at the beginning saying we are not endorsing?

hellohuanlin replied:
iirc there's something called 'flutter favorite program'? maybe only list things that belong to that?

leighajarett replied:
Yes but it hasn't run in a few years so newer packages aren't on there, I'm working on that separately but I think that would be a blocker for now

sfshaza2 commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+the platform code can be written in Objective-C or Swift. 
+
+In your Dart code, you use the plugin's Dart API and 
+behind the scenes the native plugin code will directly call 
```
```suggestion
behind the scenes the native plugin code directly calls
```

atsansone replied:
**question:** Can the native plugin code indirectly call the system library? If not, remove _directly_ because it doesn't add any value.

sfshaza2 commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+behind the scenes the native plugin code will directly call 
+the system library being used. This means that you can write 
+the code to call the Dart API once, 
+and it will work for all the platforms 
```
```suggestion
and it works for all platforms that the plugin supports.
```

sfshaza2 commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+the system library being used. This means that you can write 
+the code to call the Dart API once, 
+and it will work for all the platforms 
+that the plugin has been written for. 
```
```suggestion
```

sfshaza2 commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+the dependency to your [`pubspec.yaml`][] file. You can then import 
+the package in your Dart file. 
+
+Occasionally, you may also need to change app settings or 
+initialization logic- which will be explained in the package's 
```
```suggestion
Occasionally, you might also need to change app settings or 
initialization logic- which should be explained in the package's 
```

sfshaza2 commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+In Swift and Objective-C, you use `PhotoKit` and the `PhotosUI` framework
+to access user photos.
+
+In Flutter, you can use the [`image_picker`][] plugin, 
+which supplies an `ImpagePicker` class that is used to 
+display the user's photo library.
```
```suggestion
In Flutter, you can use the [`image_picker`][] plugin 
that supplies an `ImpagePicker` class, which  
displays the user's photo library.
```

sfshaza2 commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+
+### Accessing the camera
+In SwiftUI and UIKit, to allow your app's users to take a photo, 
+you use the `UIImagePickerController` class, 
```
```suggestion
use the `UIImagePickerController` class
```

sfshaza2 commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+
+In Flutter, you can use the [`pay`][] plugin to 
+add Google Pay payments on Android
+ and Apple pay payments on iOS.
```
An important point.

hellohuanlin replied:
since it's 3P plugin, when we endorse it, i think we need to be careful that 
1. this plugin actually works; 
2. it's good quality; 
3. it's not abandoned and actively being maintained; 
4. maybe other criteria?

leighajarett replied:
Yeah I agree, I'm adding a sentence at the top saying we're not endorsing, but I did check to make sure all of these plugins are highly rated on pub

sfshaza2 commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+
+In Flutter, you can leverage Google's ML Kit with the [`google_ml_kit`][] plugin. 
+This supports various features like text recognition, 
+face detection, image labeling, landmark recognition and barcode scanning.
```
```suggestion
face detection, image labeling, landmark recognition, and barcode scanning.
```

sfshaza2 commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+for accessing weather conditions and forecasts. 
+
+In Flutter, you can use the [`weather`][] package, 
+which uses the [OpenWeatherMap API]. 
```
```suggestion
which uses the [OpenWeatherMap API][]. 
```

sfshaza2 reviewed (commented):
A few nits. You know, it strikes me that this page would be well suited to be a 3-column table with a "Swift/Objective-C column, a Flutter column, and a Comments column. It would greatly reduce a lot of boilerplate text.

atsansone commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+### Flutter plugin overview
+When developing an app with Flutter, you can interact 
+with system libraries using plugins. 
+Plugins are Dart libraries that contain platform-specific code. 
```
**suggestion:** Use active voice.
```suggestion
Dart calls libraries that contain platform-specific code _plugins_. 
```

atsansone commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+When developing an app with Flutter, you can interact 
+with system libraries using plugins. 
+Plugins are Dart libraries that contain platform-specific code. 
+In the case of an iOS app, 
+the platform code can be written in Objective-C or Swift. 
```
**suggestion:** Active voice. Someone performs actions in these types of docs. It doesn't happen any other way.

```suggestion
With an iOS app, you can write the code in Objective-C or Swift. 
```

atsansone commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+and it will work for all the platforms 
+that the plugin has been written for. 
+
+To learn more about plugins see [Using packages][]. 
```
**nit:** Add a comma.
```suggestion
To learn more about plugins, see [Using packages][]. 
```

atsansone commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+that the plugin has been written for. 
+
+To learn more about plugins see [Using packages][]. 
+This document links out to some commonly used plugins, 
+and their examples, but there are thousands more available on [pub.dev][]. 
+And, if you can't find what you're looking for, you can create your own. 
```
**suggestion:** Tighten up this language.
```suggestion
Though this page links to some popular plugins,
you can find thousands more on [pub.dev][]. 
If you can't find a package that meets your need, 
you can create your own. 
```

leighajarett replied:
I want to make it clear that there are examples on pub.dev as well

atsansone commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+And, if you can't find what you're looking for, you can create your own. 
+
+### Adding a plugin to your project
+When using Apple frameworks within your native project, 
+you usually just need to import it into your Swift or Objective-C file. 
```
**suggestion:** Tighten up the language.
```suggestion
When using Apple frameworks within your native project, 
import it into your Swift or Objective-C file. 
```

atsansone commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+When using Apple frameworks within your native project, 
+you usually just need to import it into your Swift or Objective-C file. 
+
+In Flutter, you must first run `flutter pub add package_name` 
+in the terminal, from the root of your project. This adds 
+the dependency to your [`pubspec.yaml`][] file. You can then import 
+the package in your Dart file. 
```
**suggestion:** This part rambles some. Tighten it up.
```suggestion
Flutter requires you to run `flutter pub add package_name` 
from the root of your project. This adds 
the dependency to your [`pubspec.yaml`][] file.
After you add the dependency, `import` the package
in your Dart file. 
```

atsansone commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+the dependency to your [`pubspec.yaml`][] file. You can then import 
+the package in your Dart file. 
+
+Occasionally, you may also need to change app settings or 
```
**question:** Can you be more specific here? Are there explicit cases when the dev would need to change the settings?

leighajarett replied:
It's dependent on the plugin. See an example here - https://pub.dev/packages/image_picker#ios

The dev needs to change settings by adding keys to info.plist

atsansone commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+---
```
**suggestion:** If this is a series of comparisons, use subheads for each language, as in:
```markdown
#### Apple Framework
...
#### Flutter
...
```

This would read more smoothly than `In native code, do...` then `In Flutter, do...`

leighajarett replied:
Will hold off on this in case we turn it into a table

atsansone reviewed (changes requested):
@leighajarett : I have some suggestions here. After you revise, I'll have another look.

leighajarett commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+the package in your Dart file. 
+
+Occasionally, you may also need to change app settings or 
+initialization logic- which will be explained in the package's 
```
Yes exactly

leighajarett commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+description: Learn how to apply knowledge of Apple's APIs and frameworks when building Flutter apps.
+---
+
+This document is for iOS developers looking to apply
```
That sounds good to me!

leighajarett commented:
I like the idea of using a table for this doc, is there an example in our docs that you could send over so I can adapt it? 

Also, I want to be clear that this is not a doc for 1P plugins. It references plugins that are also 3P. I only included plugins that are widely used by the community. But I am happy to include a sentence in this doc somewhere that says something like "some of the plugins referenced in this doc are built and maintained by members of the Flutter community" to make it clear (or some language are the idea that this is not an endorsement). Let me know what you think!

@sfshaza2 @atsansone

hellohuanlin commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -0,0 +1,200 @@
+---
+title: Leveraging Apple's System Libraries
+description: Learn how to apply knowledge of Apple's APIs and frameworks when building Flutter apps.
```
Hmmmm, both "apply knowledge of apple's API when building flutter apps" and "leverage apple's API when building flutter apps" sounds like "writing code that calls apple api", which is not what it is. 

Maybe something like "Learn about various flutter plugins that offers equivalent functionalities to Apple's framework"? may need rephrasing

leighajarett replied:
Added!

hellohuanlin commented:
we may also wanna include database, which is pretty common: https://docs.flutter.dev/cookbook/persistence/sqlite

can mention native uses either `CoreData` or `sqlite`, and flutter use [this plugin](https://pub.dev/packages/sqflite)

leighajarett commented:
I've changed some of the wording based on feedback around the title / description. I also added the "persisting data" piece (following @hellohuanlin's input on SQLite) and added a sentence saying we're not endorsing packages listed (feel free to propose a better way to message this). As a quick aside, I've only included 3P plugins that have high scores on pub.

Last, I also formatted it as a table, as suggested by @atsansone. But, it could use some formatting :)

hellohuanlin commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -45,139 +47,29 @@ Occasionally, you might also need to change app settings or
+| Use Case | Apple Framework or Class | Flutter Plugin |
+| -------- | --------------- | -------------- |
+| Accessing the photo library | `PhotoKit` (`Photos` and `PhotosUI ` frameworks) | [`image_picker`][] |
+| Accessing the camera | `UIImagePickerController` class with the `.camera` `sourceType` | [`image_picker`][] |
```
`UIImagePickerController` can also be used for accessing photo library. maybe also add this to previous row?

hellohuanlin replied:
maybe also add a row 
```
| More advanced camera features | AVFoundation | camera plugin |
```

leighajarett replied:
Added, thanks!

hellohuanlin commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -45,139 +47,29 @@ Occasionally, you might also need to change app settings or
+| Accessing the camera | `UIImagePickerController` class with the `.camera` `sourceType` | [`image_picker`][] |
+| In-app purchases | `StoreKit` | [`in_app_purchase`][], which supports both Google Play Store on Android and Apple App Store on iOS |
+| Payment processing | `PassKit` | [`pay`][], which adds Google Pay payments on Android and Apple Pay payments on iOS. |
+| Push notifications | `UserNotifications` | [`firebase_messaging`][], which uses Firebase Cloud Messaging and integrates with APNs| 
```
APNS

hellohuanlin commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -45,139 +47,29 @@ Occasionally, you might also need to change app settings or
+| Embedding maps | `MapKit` | [`google_maps_flutter`][] |
+| Network requests| `URLSession` class | [`http`][] |
+| Store key-values | `@AppStorage` property wrapper or `NSUserDefaults` class| [`shared_preferences`][] |
+| Persisting data | `CoreData` | [`sqflite`][] |
```
nit: 
```
| Database | CoreData or sqlite | sqflite |
```

leighajarett replied:
Added

hellohuanlin commented on src/development/platform-integration/ios/apple-frameworks.md:
```diff
@@ -45,139 +47,29 @@ Occasionally, you might also need to change app settings or
+| Accessing sensor data (e.g. accelerometer, gyroscope) | `CoreMotion`  | [`sensors_plus`][] |
+| Embedding maps | `MapKit` | [`google_maps_flutter`][] |
+| Network requests| `URLSession` class | [`http`][] |
+| Store key-values | `@AppStorage` property wrapper or `NSUserDefaults` class| [`shared_preferences`][] |
```
Nit: `@AppStorage` (in SwiftUI)

leighajarett replied:
Added

hellohuanlin reviewed (commented):
Table for the win

leighajarett commented:
Thanks for the formatting, looks good to me! Is there anyone else that needs to review @atsansone?
